src/Products/Viewer/Cache/FileViewerCache.cs

```diff
@@ -0,0 +1,210 @@
+        {
+            string cacheFilePath = this.GetCacheFilePath(key);
+
+            if (File.Exists(cacheFilePath))
+            {
+                value = typeof(T) == typeof(Stream)
+                    ? (T)this.ReadStream(cacheFilePath)
+                    : (T)this.Deserialize(cacheFilePath);
+
+                return true;
+            }
+
+            value = default(T);
+            return false;
+        }
+
+        private object ReadStream(string cacheFilePath)
+        {
+            return this.GetStream(cacheFilePath, FileMode.Open, FileAccess.Read, FileShare.Read);
+        }
+
+        private object Deserialize(string cachePath)
+        {
+            object data;
+            using (FileStream stream = this.GetStream(cachePath, FileMode.Open, FileAccess.Read, FileShare.Read))
+            {
+                BinaryFormatter formatter = new BinaryFormatter();
+                formatter.Binder = new IgnoreAssemblyVersionSerializationBinder();
+
+                try
+                {
+                    data = formatter.Deserialize(stream);
+                }
+                catch (SerializationException)
+                {
+                    data = null;
+                }
+            }
+
+            return data;
+        }
+
+        public string GetCacheFilePath(string key)
+        {
+            string folderPath = Path.Combine(this.CachePath, this.CacheSubFolder);
+            string filePath = Path.Combine(folderPath, key);
+
+            if (!Directory.Exists(folderPath))
+            {
+                Directory.CreateDirectory(folderPath);
+            }
+
+            return filePath;
+        }
+
+        public bool Contains(string key)
+        {
+            string file = Path.Combine(this.CachePath, this.CacheSubFolder, key);
+            return File.Exists(file);
+        }
+
+        private FileStream GetStream(string path, FileMode mode, FileAccess access, FileShare share)
+        {
+            FileStream stream = null;
+            TimeSpan interval = new TimeSpan(0, 0, 0, 0, 50);
+            TimeSpan totalTime = new TimeSpan();
+
+            while (stream == null)
+            {
+                try
+                {
+                    stream = File.Open(path, mode, access, share);
+                }
+                catch (IOException)
+                {
+                    Thread.Sleep(interval);
+                    totalTime += interval;
+
+                    if (this.waitTimeout.Ticks != 0 && totalTime > this.waitTimeout)
+                    {
+                        throw;
+                    }
+                }
+            }
+
+            return stream;
+        }
+
+        private void CopyStream(Stream src, Stream dst)
+        {
+            const int bufferSize = 81920; //NOTE: taken from System.IO
+            byte[] buffer = new byte[bufferSize];
+            int read;
+            while ((read = src.Read(buffer, 0, buffer.Length)) != 0)
```

Comment:
Codacy found an issue: [Add curly braces around the nested statement(s) in this 'while' block.](https://app.codacy.com/manual/GroupDocs/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5933706)